### PR TITLE
syncer(dm): add the function GetTimezoneByStatusVars()

### PR DESCRIPTION
**Pull request by lyzx2001**
### What problem does this PR solve?

Issue Number: ref #5066 

### What is changed and how it works?
The function GetTimezoneByStatusVars() is added to retrieve the upstream time zone for "ALTER TABLE ADD COLUMN x DATETIME", but not for "ALTER TABLE ADD COLUMN x TIMESTAMP", since there is no time zone information for the latter one. This function will be used in further fixing issue #5066.

It uses the function statusVarsToKV() to retrieve the time zone and convert it to string. If the result is "SYSTEM", replace it with Syncer.upstreamTZ, in order to guarantee that the absolute time zone of upstream will be synchronized to downstream.

Also, I have added the unit test for this function. The input data for this unit test were retrieved using the function Dump() in go-mysql.

#### Tests 

 - Unit test
 

##### Will it cause performance regression or break compatibility?
No.

##### Do you need to update user documentation, design documentation or monitoring documentation?
No.

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/pkg/binlog/event/util.go [hunk 1]
```diff
@@ -408,6 +408,27 @@ func GetCharsetCodecByStatusVars(statusVars []byte) (encoding.Encoding, error) {
...
 
+// GetTimezoneByStatusVars returns the timezone of upstream for "datetime" type alter table commands
+// For "ALTER TABLE ADD x datetime", not for "ALTER TABLE ADD x timestamp".
+func GetTimezoneByStatusVars(statusVars []byte, upstreamTZStr string) (string, error) {
+	vars, err := statusVarsToKV(statusVars)
+	b, ok := vars[QTimeZoneCode]
+
+	if !ok {
+		if err == nil {
+			// only happen when this is a dummy event generated by DM
```
Comment (lance6716) [inline dm/pkg/binlog/event/util.go:419]: this comment should be updated, as in the comment of function some DDL will also not write timezone to status vars

File: dm/pkg/binlog/event/util_test.go [hunk 2]
```diff
@@ -133,3 +139,38 @@ func (t *testUtilSuite) TestStatusVarsToKV(c *C) {
...
+			nil,
+		},
+	}
+
+	for _, t := range testCases {
+		var upstreamTZStr string // to stimulate Syncer.upstreamTZ
+		upstreamTZStr = "+0:00"
+		vars, err := GetTimezoneByStatusVars(t.input, upstreamTZStr)
+		if t.err != nil {
+			c.Assert(err.Error(), Equals, t.err.Error())
```
Comment (lance6716) [inline dm/pkg/binlog/event/util_test.go:170]: please also use testify package in this file. Also change the old tests.

File: dm/tests/many_tables/run.sh [hunk 3]
```diff
@@ -16,6 +16,7 @@ function prepare_data() {
 		for j in $(seq 2); do
 			run_sql "INSERT INTO many_tables_db.t$i VALUES ($j,${j}000$j),($j,${j}001$j);" $MYSQL_PORT1 $MYSQL_PASSWORD1
 		done
+		run_sql "INSERT INTO many_tables_db.t$i VALUES (9, 90009);" $MYSQL_PORT1 $MYSQL_PASSWORD1
```
Comment (lance6716) [inline dm/tests/many_tables/run.sh:19]: changes of this file should not be included

File: dm/pkg/binlog/event/util.go [hunk 4]
```diff
@@ -408,6 +408,27 @@ func GetCharsetCodecByStatusVars(statusVars []byte) (encoding.Encoding, error) {
...
+			err = fmt.Errorf("Timezone not found in status_vars %v", statusVars)
+		}
+		return "", err
+	}
+
+	s := string(b[1:])
```
Comment (D3Hunter) [inline dm/pkg/binlog/event/util.go:425]: plz leave a comment about the structure of this value

File: dm/pkg/binlog/event/util.go [hunk 5]
```diff
@@ -408,6 +408,27 @@ func GetCharsetCodecByStatusVars(statusVars []byte) (encoding.Encoding, error) {
...
 }
 
+// GetTimezoneByStatusVars returns the timezone of upstream for "datetime" type alter table commands
+// For "ALTER TABLE ADD x datetime", not for "ALTER TABLE ADD x timestamp".
+func GetTimezoneByStatusVars(statusVars []byte, upstreamTZStr string) (string, error) {
+	vars, err := statusVarsToKV(statusVars)
+	b, ok := vars[QTimeZoneCode]
+
+	if !ok {
+		if err == nil {
```
Comment (D3Hunter) [inline dm/pkg/binlog/event/util.go:418]: this part of code is quite anti-pattern, seems imitate logic in getSQLMode(GetServerCollationByStatusVars too). it's introduced in https://github.com/pingcap/dm/pull/1176/files, to fix some undocumented var which may causes syncer stop. 

we should leave a comment to explain why we didn't check error first. or better to refactor this part of code, we don't care about undocumented vars anyway, it should not return error.

Comment (D3Hunter) [reply dm/pkg/binlog/event/util.go:418]: fired an issue https://github.com/pingcap/tiflow/issues/6121

Comment (lyzx2001) [reply dm/pkg/binlog/event/util.go]: I have modified this part of the code. Please take a look at whether it corresponds to what you meant.

File: dm/tests/many_tables/run.sh (hunk 3 above)
Comment (lyzx2001) [reply dm/tests/many_tables/run.sh]: Have removed this part of the changes.

File: dm/pkg/binlog/event/util_test.go (hunk 2 above)
Comment (lyzx2001) [reply dm/pkg/binlog/event/util_test.go]: Done.

File: dm/pkg/binlog/event/util.go (hunk 4 above)
Comment (lyzx2001) [reply dm/pkg/binlog/event/util.go:423]: Done.

File: dm/pkg/binlog/event/util.go [hunk 6]
```diff
@@ -408,6 +408,26 @@ func GetCharsetCodecByStatusVars(statusVars []byte) (encoding.Encoding, error) {
 	}
 }
 
+// GetTimezoneByStatusVars returns the timezone of upstream for "datetime" type alter table commands.
+// For "ALTER TABLE ADD COLUMN x DATETIME", not for "ALTER TABLE ADD COLUMN x TIMESTAMP".
+func GetTimezoneByStatusVars(statusVars []byte, upstreamTZStr string) (string, error) {
+	vars, err := statusVarsToKV(statusVars)
+	b, _ := vars[QTimeZoneCode]
```
Comment (D3Hunter) [inline dm/pkg/binlog/event/util.go:415]: didn't check whether `QTimeZoneCode` exist. previous impl is ok, but maybe need add some comments about why we didn't check error first as usual.

as of `undocumented vars`, it's related to `statusVarsToKV`, we can refactor it later.

File: dm/pkg/binlog/event/util_test.go [hunk 7]
```diff
@@ -17,24 +17,27 @@
 package event
 
 import (
+	"github.com/stretchr/testify/require"
```
Comment (lance6716) [inline dm/pkg/binlog/event/util_test.go:20]: configurate the IDE to group and sort the import, like

https://www.jetbrains.com/help/go/creating-and-optimizing-imports.html#use-the-goimports-local-grouping

File: dm/pkg/binlog/event/util_test.go [hunk 8]
```diff
@@ -123,13 +126,50 @@ func (t *testUtilSuite) TestStatusVarsToKV(c *C) {
...
-	for _, t := range testCases {
-		vars, err := statusVarsToKV(t.input)
-		if t.err != nil {
-			c.Assert(err.Error(), Equals, t.err.Error())
+	for _, test := range testCases {
+		vars, err := statusVarsToKV(test.input)
+		if test.err != nil {
+			require.Equal(t, test.err.Error(), err.Error())
+		} else {
+			require.Nil(t, err)
```
Comment (lance6716) [inline dm/pkg/binlog/event/util_test.go:134]:
```suggestion
			require.NoError(t, err)
```

File: dm/pkg/binlog/event/util_test.go [hunk 9]
```diff
@@ -123,13 +126,50 @@ func (t *testUtilSuite) TestStatusVarsToKV(c *C) {
...
+		}
+		require.Equal(t, vars, test.output)
+
```
Comment (lance6716) [inline dm/pkg/binlog/event/util_test.go:137]:
```suggestion
```

File: dm/pkg/binlog/event/util_test.go [hunk 10]
```diff
@@ -123,13 +126,50 @@ func (t *testUtilSuite) TestStatusVarsToKV(c *C) {
...
+	for _, test := range testCases {
+		var upstreamTZStr string // to simulate Syncer.upstreamTZ
+		upstreamTZStr = "+0:00"
+		vars, err := GetTimezoneByStatusVars(test.input, upstreamTZStr)
+		if test.err != nil {
+			require.Equal(t, test.err.Error(), err.Error())
 		} else {
-			c.Assert(err, IsNil)
+			require.Nil(t, err)
```
Comment (lance6716) [inline dm/pkg/binlog/event/util_test.go:171]:
```suggestion
			require.NoError(t, err)
```

File: dm/pkg/binlog/event/util_test.go [hunk 11]
```diff
@@ -123,13 +126,50 @@ func (t *testUtilSuite) TestStatusVarsToKV(c *C) {
...
 		}
-		c.Assert(vars, DeepEquals, t.output)
+		require.Equal(t, vars, test.output)
```
Comment (lance6716) [inline dm/pkg/binlog/event/util_test.go:173]:
```suggestion
		require.Equal(t, test.output, vars)
```

The first is expected and second is got. Please check parameter names.

File: dm/pkg/binlog/event/util.go (hunk 6 above)
Comment (lyzx2001) [reply dm/pkg/binlog/event/util.go]: Understand. Added the comment.

**Review by lance6716 — APPROVED** [review]

**Comment by lance6716** [comment]
ptal @D3Hunter

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**